cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp

```diff
@@ -0,0 +1,448 @@
+                                  const ndview<Float, 2>& data,
+                                  ndview<Float, 1>& probabilities,
+                                  const event_vector& deps) {
+    auto fill_event = fill<Float>(q, probabilities, Float(1), {});
+    using oneapi::dal::backend::operator+;
+
+    const std::int64_t n = data.get_dimension(0);
+
+    auto param_arr = ndarray<Float, 1>::wrap(parameters.get_data(), 1);
+    Float w0 = param_arr.to_host(q, deps).at(0); // Poor perfomance
+
+    auto event = gemv(q,
+                      data,
+                      parameters.get_slice(1, parameters.get_dimension(0)),
+                      probabilities,
+                      Float(1),
+                      w0,
+                      { fill_event });
+    auto prob_ptr = probabilities.get_mutable_data();
+
+    return q.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(event);
+        const auto range = make_range_1d(n);
+        cgh.parallel_for(range, [=](sycl::id<1> idx) {
+            prob_ptr[idx] = 1 / (1 + sycl::exp(-prob_ptr[idx]));
+        });
+    });
+    // return event;
+}
+
+template <typename Float>
+sycl::event compute_logloss(sycl::queue& q,
+                            const ndview<Float, 1>& parameters,
+                            const ndview<Float, 2>& data,
+                            const ndview<std::int32_t, 1>& labels,
+                            const ndview<Float, 1>& probabilities,
+                            ndview<Float, 1>& out,
+                            Float L1,
+                            Float L2,
+                            const event_vector& deps) {
+    const std::int64_t n = data.get_dimension(0);
+    const std::int64_t p = data.get_dimension(1);
+    ONEDAL_ASSERT(parameters.get_dimension(0) == p + 1);
+    ONEDAL_ASSERT(labels.get_dimension(0) == n);
+    ONEDAL_ASSERT(labels.has_data());
+    ONEDAL_ASSERT(parameters.has_data());
+    ONEDAL_ASSERT(data.has_data());
+
+    auto labels_ptr = labels.get_data();
+    auto prob_ptr = probabilities.get_data();
+
+    auto out_ptr = out.get_mutable_data();
+
+    auto loss_event = q.submit([&](sycl::handler& cgh) {
+        const auto range = make_range_1d(n);
+        using oneapi::dal::backend::operator+;
+        using sycl::reduction;
+
+        cgh.depends_on(deps);
+
+        auto sumReduction = reduction(out_ptr, sycl::plus<>());
+
+        cgh.parallel_for(range, sumReduction, [=](sycl::id<1> idx, auto& sum) {
+            const Float prob = prob_ptr[idx];
+            const std::int32_t label = labels_ptr[idx];
+            sum += -label * sycl::log(prob) - (1 - label) * sycl::log(1 - prob);
+        });
+    });
+
+    auto [out_reg, out_reg_e] = ndarray<Float, 1>::zeros(q, { 1 }, sycl::usm::alloc::device);
+    auto reg_ptr = out_reg.get_mutable_data();
+    event_vector vector_out_reg = { out_reg_e };
+
+    auto param_ptr = parameters.get_data();
```

Comment:
Please use `const auto* const` for pointers to constant data and `auto* const` for pointers to modifiable data